### DISCUSSION TITLE
Do we have a restore method available?

### DISCUSSION THREAD

**Discussion by cengiz-io** [rgerganov/gen5fw#2] [General]
Hello @rgerganov

I just saw your blog post and it certainly sparked some ideas on my mind.

However, since I use my car almost daily, I'd like to know if we have a restoration method available if one decides to flash the bootloader / littlekernel / partition.

I'm have embedded linux experience, particularly on ARM Cortex-A chips (and U-Boot of course) but I would like to learn more about the possible permanent damage we can do before moving further.

Thanks a lot!

**Comment by rgerganov** [comment]
Hi @cengiz-io 
The method being described here is what they call "recovery" method. It is supposed to work as long as the `lk.rom` is not damaged. Even if you mess up the root filesystem in `system.ext4`, you can always recover the original one from the stock firmware using this method. So if you don't make any changes to `lk.rom`, the recovery method should be always available.

**Comment by St0nehead** [comment]
Can I use this toolkit to modify the Gen5w firmware? Or is it intended only for Gen5? 
At the moment, I only need to change the firmware from Korean to European.
Thanks in advance

**Reply by rgerganov** [reply]
> Can I use this toolkit to modify the Gen5w firmware? Or is it intended only for Gen5? At the moment, I only need to change the firmware from Korean to European. Thanks in advance

No, Gen5w is different from Gen5.

**Comment by adelie93** [comment]
![s-l1600](https://user-images.githubusercontent.com/109261315/236949589-423c3cca-fee5-4496-834f-b211b955ce07.jpg)

I have the above head unit (Kia Niro 2019 EV). The one pictured is not mine, just a photo I found on the internet, mine is in the car of course.

Any suggestions how I can get into the bootloader menu? As you can see I do not have a "Map" button. I tried holding "Power" plus "Home" as suggested elsewhere while starting the car, but nothing happened. Am I doing it wrong or do I need to find some other buttons?

**Reply by rgerganov** [reply]
> ![s-l1600](https://user-images.githubusercontent.com/109261315/236949589-423c3cca-fee5-4496-834f-b211b955ce07.jpg)
> 
> I have the above head unit (Kia Niro 2019 EV). The one pictured is not mine, just a photo I found on the internet, mine is in the car of course.
> 
> Any suggestions how I can get into the bootloader menu? As you can see I do not have a "Map" button. I tried holding "Power" plus "Home" as suggested elsewhere while starting the car, but nothing happened. Am I doing it wrong or do I need to find some other buttons?

You can try "Power"+"Media" or "Power"+"Setup"

**Reply by adelie93** [reply]
Some good news and some bad news.

Good news: "Power" + "Home" did actually work. The problem is that the head unit was not actually starting from a cold boot, so the special keys did nothing. When I pulled the fuse and let it boot from cold, the keys worked!

Bad news:
I followed the directions 100%, only trying to enable ADB. The flash did work, but the system behaves very oddly now.  
All of the "Settings" have gone wrong. All text is replaced with "Manage applications"!

![Settings_Menu](https://github.com/rgerganov/gen5fw/assets/109261315/03581550-4017-4af1-8f6b-dd931f9c9034)

Most frighteningly, none of the Settings buttons actually respond. I can see from the icons which should be which, but when I press them, nothing happens. So I am unable to reach the screen with the "Upgrade" button to try the "intended" flashing method to recover.

I am open to suggestions!

**Reply by adelie93** [reply]
I tried re-flashing it with a completely unmodified image, same result.

Also after re-flashing, all user data is gone. It gives a message about newly enrolling in UVO.
My guess is that maybe your `partition.dat` file is different from what it is supposed to be for my device. But I am not sure how to determine that for sure, or how to fix it...

**Reply by rgerganov** [reply]
Are you sure that you are using a firmware image from the original region (e.g. Middle East, EU, CA)? We have seen this happening when flashing an image from another region, for example Middle East over CA.

**Reply by adelie93** [reply]
Yes, it's the same region. It's the exact same firmware image, in fact.
I first used the firmware image that I downloaded to do a normal upgrade. This worked fine.
Then I tried the `security_force` upgrade as outlined here using the `system.ext4` image extracted from the same firmware I'd just installed.

**Reply by adelie93** [reply]
Looking inside the `update.zip`, there is a file that looks like what is supposed to happen on a "normal" upgrade: `META-INF/com/google/android/updater-script`.

I won't copy the whole thing here, but here is the end:
```
ui_print("Write splash.img ...");
package_extract_file("splash.img", "/upgrade/splash.img");
write_raw_image("/upgrade/splash.img", "splash");
delete("/upgrade/splash.img");

ui_print("\nsystem partition formatting...");
format("ext4","EMMC","/dev/block/mmcblk0p2","452984832","system");

ui_print("\nSystem partition writing...");
package_extract_file("system.ext4", "/tmp/system.ext4");
write_raw_image("/tmp/system.ext4", "system");
delete("/tmp/system.ext4");

ui_print("Write snapshot.img ...");
write_sparse_image_no_zip("/upgrade/snapshot.sparse.img", "/dev/block/mmcblk0p10");

ui_print("\nD-Audio system update finished\n");
```

Looks to me like part of the update process is to write the `snapshot.sparse.img` file to the proper partition. Could this be causing the issue?
Is there a way to write this extra image from within the `security_force` framework?

**Reply by adelie93** [reply]
I wrote the `snapshot.sparse.img` to `blk0p10` and it just boot looped. So that isn't a good idea. 😄 
Re-doing the `security_force` install gets me back to where things were before. Basic functions work but many items still show `Manage applications`.
